python-frontend/src/main/java/org/sonar/python/semantic/v2/typeshed/ProtoUtils.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.semantic.v2.typeshed;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.sonar.plugins.python.api.PythonVersionUtils;
+import org.sonar.python.index.AmbiguousDescriptor;
+import org.sonar.python.index.Descriptor;
+
+public class ProtoUtils {
+  private ProtoUtils() {
+    // utility class
+  }
+
+  static Map<String, Descriptor> disambiguateByName(Stream<Stream<? extends Descriptor>> input) {
+    return input.flatMap(i -> i)
+      .collect(Collectors.groupingBy(
+        Descriptor::name, Collectors.collectingAndThen(
+          Collectors.toSet(),
+          ProtoUtils::disambiguateSymbolsWithSameName)));
+  }
+
+  private static Descriptor disambiguateSymbolsWithSameName(Set<Descriptor> descriptors) {
+    if (descriptors.size() > 1) {
+      return AmbiguousDescriptor.create(descriptors);
+    }
+    return descriptors.iterator().next();
+  }
+
+  static boolean isValidForPythonVersion(List<String> validForPythonVersions, Set<String> supportedPythonVersions) {
```

Comment:
You mean because we're creating this intersection set for all tested symbol? Do you think this is significant enough that we should rethink the approach already?

Reply:
This is one of the problem. As discussed, we should create a new ticket to hoist the TypeShedDescriptorsProvider up to the ProjectLevelSymbolTable